torch/nn/_functions/rnn.py

```diff
@@ -74,7 +81,11 @@ def forward(input, hidden, weight):
         next_hidden = []
 
         if lstm:
-            hidden = list(zip(*hidden))
+            # NB: Be sure to use the generator version of zip here in
+            # Python 2, since the evaluation order differs (Python 2
+            # evaluates all of the first argument, and then evaluates
+            # all of the second argument; generator version interleaves.)
```

Comment:
Uh how does the argument evaluation order matter here?

Reply:
Because the index here is a *tensor indexing* expression, which is an autograd operation which shows up in the trace.

Reply:
Uh it sounds more like sth that should be fixed in tracer tests. Both are valid traces, so there's no need to enforce full determinism in all library code.